somatic_llr_filter.py

```diff
@@ -0,0 +1,303 @@
+                       normal_ref, normal_var, tumor_ref, tumor_var))
+    elif call == "NotSomatic":
+        return(calc_llr(error_expect, error_expect,
+                       normal_ref, normal_var, tumor_ref, tumor_var))
+    elif call == "Reference":
+        return(calc_llr(error_rate, error_rate,
+                       normal_ref, normal_var, tumor_ref, tumor_var))
+    elif call == "Somatic":
+        return(calc_llr(error_rate + (tumor_freq / tumor_purity * normal_contamination_rate), tumor_freq,
+                       normal_ref, normal_var, tumor_ref, tumor_var))
+    raise Exception('Call "' + call + '" is not a recognized type')
+
+
+def make_call(normal_ref, normal_var, tumor_ref, tumor_var, error_rate, heterozygous_expect,
+              homozygous_expect, tumor_freq, tumor_purity, normal_contamination_rate, error_expect):
+    #Want to test several models and pick the most likely one
+    #Reference: normal expectation is 0.001 and tumor expectation is 0.01
+    #Germline Het: normal expectation is 0.5 and tumor expectation is 0.5
+    #Somatic Het: normal expectation is 0.001 and tumor expectation in 0.5
+    #Germline Homozygote: normal expectation is 0.999 and tumor expectation is 0.999
+    #LOH (variant): germline is 0.5 and tumor is 0.999
+    #LOH (ref): germline is 0.5 and tumor is 0.001
+
+    marginal_probability = None
+    max_llr = None
+    max2_llr = None
+    max_call = None
+    max2_call = None
+    llr = 0
+
+    if(tumor_var == 0): #special case that chokes some of the log code.  LLR is always zero if we have no supp reads
+        return([0,"Reference"])
+
+    call_types = ["Germline_het", "Germline_hom", "LOH_ref", "LOH_variant", "NotSomatic", "Reference", "Somatic"];
+    for call in call_types:
+        llr = get_llr(call, normal_ref, normal_var, tumor_ref, tumor_var, error_rate, heterozygous_expect,
+                      homozygous_expect, tumor_freq, tumor_purity, normal_contamination_rate, error_expect)
+        if marginal_probability is None:
+            marginal_probability=llr
+        elif llr > marginal_probability:
+            marginal_probability = llr + math.log(1 + math.exp(marginal_probability-llr))
+        else:
+            marginal_probability = marginal_probability + math.log(1 + math.exp(llr-marginal_probability))
+
+        max_llr_def = (max_llr is not None)
+        max2_llr_def = (max2_llr is not None)
+
+        if not (max_llr_def and (llr < max_llr)):
+            #save the second-place, store the new best
+            max2_call = max_call
+            max2_llr = max_llr
+            max_llr = llr
+            max_call = call
+        elif not (max2_llr_def and (llr < max2_llr)):
+            max2_call = call
+            max2_llr = llr
+    #end for
+
+    #get the final llr
+    if max_llr is not None:
+        llr = max_llr-max2_llr
+    else:
+        llr = 0
+
+    #get the final call
+    if max_call is None:
+        max_call = "-"
+
+    return [llr,max_call]
```

Comment:
You can also return this as `(llr, max_call)` and when you call this method `(llr, max_call) = make_call(...` That way you don't have to access array elements on lines 291 and 296.

Reply:
yeah, that is cleaner - implemented